cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1272,715 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+ * iterations in addition to helping keep the state about what batches have been copied so far
+ * and what are the sizes (in bytes) of each iteration.
+ */
+struct chunk_iteration_state {
+  chunk_iteration_state(rmm::device_uvector<dst_buf_info> _d_batched_dst_buf_info,
+                        rmm::device_uvector<offset_type> _d_batch_offsets,
+                        std::vector<std::size_t>&& _h_num_buffs_per_iteration,
+                        std::vector<std::size_t>&& _h_size_of_buffs_per_iteration,
+                        std::size_t total_size)
+    : num_iterations(_h_num_buffs_per_iteration.size()),
+      current_iteration(0),
+      starting_batch(0),
+      d_batched_dst_buf_info(std::move(_d_batched_dst_buf_info)),
+      d_batch_offsets(std::move(_d_batch_offsets)),
+      h_num_buffs_per_iteration(std::move(_h_num_buffs_per_iteration)),
+      h_size_of_buffs_per_iteration(std::move(_h_size_of_buffs_per_iteration)),
+      total_size(total_size)
+  {
+  }
+
+  /**
+   * @brief As of the time of the call, return the starting 1MB batch index, and the
+   * number of batches to copy.
+   *
+   * @return the current iteration's starting_batch and batch count as a pair
+   */
+  std::pair<std::size_t, std::size_t> get_current_starting_index_and_buff_count() const
+  {
+    CUDF_EXPECTS(current_iteration < num_iterations,
+                 "current_iteration cannot exceed num_iterations");
+    auto count_for_current = h_num_buffs_per_iteration[current_iteration];
+    return std::make_pair(starting_batch, count_for_current);
+  }
 
-  // HtoD src and dest buffers
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    d_src_bufs, h_src_bufs, src_bufs_size + dst_bufs_size, cudaMemcpyDefault, stream.value()));
+  std::size_t advance_iteration()
+  {
+    CUDF_EXPECTS(current_iteration < num_iterations,
+                 "current_iteration cannot exceed num_iterations");
+    std::size_t bytes_copied = h_size_of_buffs_per_iteration[current_iteration];
+    starting_batch += h_num_buffs_per_iteration[current_iteration];
+    ++current_iteration;
+    return bytes_copied;
+  }
 
-  // perform the copy.
-  copy_data(num_bufs, num_src_bufs, d_src_bufs, d_dst_bufs, d_dst_buf_info, stream);
+  bool has_more_copies() const { return current_iteration < num_iterations; }
 
-  // DtoH dst info (to retrieve null counts)
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    h_dst_buf_info, d_dst_buf_info, dst_buf_info_size, cudaMemcpyDefault, stream.value()));
+  rmm::device_uvector<dst_buf_info> d_batched_dst_buf_info;
+  rmm::device_uvector<offset_type> const d_batch_offsets;
+  std::size_t const total_size;
+  int const num_iterations;
+  int current_iteration;
 
-  stream.synchronize();
+ private:
+  std::size_t starting_batch;
+  std::vector<std::size_t> const h_num_buffs_per_iteration;
+  std::vector<std::size_t> const h_size_of_buffs_per_iteration;
+};
+
+std::unique_ptr<chunk_iteration_state> make_chunk_iteration_state(
```

Comment:
Please follow the factory function style:
```
auto chunk_iteration_state::make(....) {...}
```

Reply:
actually in this case, can we keep it separate? This function is very large, and I would rather keep `chunk_iteration_state` small so you can see it all in one page. This is an internal implementation detail, so I am hoping we can let this factory one slide.

Reply:
Okay that makes sense.

Reply:
An alternate solution here that would satisfy both requests is to define the function out of band, i.e.
```
class chunk_iteration_state {
...
static auto create(...);
...
}
static auto chunk_iteration_state::create(...) {...}
```
WDYT?